### PR TITLE
RediSearch Support

### PR DESCRIPTION
**Pull request by ofekshenawa**

### CODE REVIEW

File: search_commands.go [hunk 1]
```diff
@@ -0,0 +1,1215 @@
+package redis
+
+import (
+	"context"
+	"strings"
+)
+
+type SearchCmdable interface {
+	FT_List(ctx context.Context) *StringSliceCmd
```
Comment (chayim) [inline search_commands.go:9]: Can you ping Adriano - I know we have no other way.. but this exposes this (specific) item, even though we may want to walk it back

File: search_commands.go [hunk 2]
```diff
@@ -0,0 +1,1215 @@
...
+	FTConfigSet(ctx context.Context, option string, value interface{}) *StatusCmd
+	FTCreate(ctx context.Context, index string, options *FTCreateOptions, schema ...*SearchSchema) *StatusCmd
+	FTCursorDel(ctx context.Context, index string, cursorId int) *StatusCmd
+	FTCursorRead(ctx context.Context, index string, cursorId int, count int) *MapStringInterfaceCmd
+	FTDictAdd(ctx context.Context, dict string, term []interface{}) *IntCmd
+	FTDictDel(ctx context.Context, dict string, term []interface{}) *IntCmd
+	FTDictDump(ctx context.Context, dict string) *StringSliceCmd
+	FTDropIndex(ctx context.Context, index string) *StatusCmd
+	FTDropIndexWithArgs(ctx context.Context, index string, options *FTDropIndexOptions) *StatusCmd
+	FTExplain(ctx context.Context, index string, query string) *StringCmd
```
Comment (chayim) [inline search_commands.go:26]: Should was add explaincli and follow the panic approach we took with others?

File: search_commands.go [hunk 3]
```diff
@@ -0,0 +1,1215 @@
...
+	FTExplainWithArgs(ctx context.Context, index string, query string, options *FTExplainOptions) *StringCmd
+	FTInfo(ctx context.Context, index string) *MapStringInterfaceCmd
+	FTProfileSearch(ctx context.Context, index string, limited bool, query SearchQuery) *MapStringInterfaceCmd
```
Comment (chayim) [inline search_commands.go:29]: It's fine to have FTProfile for use cases, but let's also have a generic, that accepts an enum for SEARCH||AGGREGATE

File: search_commands.go [hunk 4]
```diff
@@ -0,0 +1,1215 @@
...
+	Score           float64
+	ScoreField      string
+	PayloadField    string
+	MaxTextFields   int
+	NoOffsets       bool
+	Temporary       int
+	NoHL            bool
+	NoFields        bool
+	NoFreqs         bool
+	StopWords       []interface{}
```
Comment (chayim) [inline search_commands.go:57]: IMHO StopWords should be an []string and we should insert based on the length of the array.

File: search_commands.go [hunk 5]
```diff
@@ -0,0 +1,1215 @@
...
+	FTSynDump(ctx context.Context, index string) *MapStringSliceInterfaceCmd
+	FTSynUpdate(ctx context.Context, index string, synGroupId interface{}, terms []interface{}) *StatusCmd
+	FTSynUpdateWithArgs(ctx context.Context, index string, synGroupId interface{}, options *FTSynUpdateOptions, terms []interface{}) *StatusCmd
+	FTTagVals(ctx context.Context, index string, field string) *StringSliceCmd
+}
+
+type FTCreateOptions struct {
+	OnHash          bool
+	OnJSON          bool
+	Prefix          []interface{}
```
Comment (chayim) [inline search_commands.go:44]: []string?
Then insert the count based on its length, while creating the string

File: search_commands.go [hunk 6]
```diff
@@ -0,0 +1,1215 @@
...
+	Sortable          bool
+	UNF               bool
+	NoStem            bool
+	NoIndex           bool
+	PhoneticMatcher   string
+	Weight            float64
+	Seperator         string
+	CaseSensitive     bool
+	WithSuffixtrie    bool
+	VectorArgs        *FTVectorArgs
```
Comment (chayim) [inline search_commands.go:74]: Not all SerachSchemas require a vector. IMHO we should revisit this collection. Maybe we have a schema per type of schemafield?

File: search_commands.go [hunk 7]
```diff
@@ -0,0 +1,1215 @@
...
+	NoOffsets       bool
+	Temporary       int
+	NoHL            bool
+	NoFields        bool
+	NoFreqs         bool
+	StopWords       []interface{}
+	SkipInitalScan  bool
+}
+
+type SearchSchema struct {
```
Comment (chayim) [inline search_commands.go:61]: I'd argue this is a FieldSchema and not a searchSchema

File: search_commands.go [hunk 8]
```diff
@@ -0,0 +1,1215 @@
...
+				}
+				if schema.VectorArgs.HNSWOptions.Epsilon > 0 {
+					hnswArgs = append(hnswArgs, "EPSILON", schema.VectorArgs.HNSWOptions.Epsilon)
+				}
+				args = append(args, len(hnswArgs))
+				args = append(args, hnswArgs...)
+			}
+		}
+		if schema.GeoShapeFieldType != "" {
+			if strings.ToUpper(schema.FieldType) != "GEOSHAPE" {
```
Comment (chayim) [inline search_commands.go:715]: GEOSHAPE should be a const.
IMHO all field types should be an enum perhaps

File: search_commands.go [hunk 9]
```diff
@@ -0,0 +1,1215 @@
...
+	args := []interface{}{"FT.DICTDEL", dict}
+	args = append(args, term...)
+	cmd := NewIntCmd(ctx, args...)
+	_ = c(ctx, cmd)
+	return cmd
+}
+
+// FTDictDump - Returns all terms in the specified dictionary.
+// The 'dict' parameter specifies the dictionary from which to return the terms.
+// For more information, please refer to the Redis documentation: [FT.DICTDUMP](https://redis.io/commands/ft.dictdump/)
```
Comment (chayim) [inline search_commands.go:799]: Does the markdown render on go docs?

File: search_commands.go [hunk 10]
```diff
@@ -0,0 +1,1215 @@
...
+	FTSpellCheckWithArgs(ctx context.Context, index string, query string, options *FTSpellCheckOptions) *MapStringInterfaceCmd
+	FTSearch(ctx context.Context, index string, query string) *Cmd
+	FTSearchWithArgs(ctx context.Context, index string, query string, options *FTSearchOptions) *Cmd
+	FTSynDump(ctx context.Context, index string) *MapStringSliceInterfaceCmd
+	FTSynUpdate(ctx context.Context, index string, synGroupId interface{}, terms []interface{}) *StatusCmd
+	FTSynUpdateWithArgs(ctx context.Context, index string, synGroupId interface{}, options *FTSynUpdateOptions, terms []interface{}) *StatusCmd
+	FTTagVals(ctx context.Context, index string, field string) *StringSliceCmd
+}
+
+type FTCreateOptions struct {
```
Comment (chayim) [inline search_commands.go:41]: Given the variety, I think NewFTIndexWith<> functions are a good idea. You can talk me out of it for initial release, but I think it has value

File: search_test.go [hunk 11]
```diff
@@ -0,0 +1,1205 @@
...
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+var _ = Describe("RediSearch commands", Label("search"), func() {
+	ctx := context.TODO()
+	var client *redis.Client
+
+	BeforeEach(func() {
+		client = redis.NewClient(&redis.Options{Addr: ":6379"})
```
Comment (chayim) [inline search_test.go:28]: Can you get the Addr from the same client we already use?

File: search_test.go [hunk 12]
```diff
@@ -0,0 +1,1205 @@
...
+		Expect(err).NotTo(HaveOccurred())
+		Expect(val).To(BeEquivalentTo("OK"))
+		WaitForIndexing(client, "txt")
+		client.HSet(ctx, "doc1", "txt", "foo baz")
+		client.HSet(ctx, "doc2", "txt", "foo bar")
+		res, err := client.FTSearchWithArgs(ctx, "txt", "foo ~bar", &redis.FTSearchOptions{WithScores: true}).Result()
+		Expect(err).NotTo(HaveOccurred())
+		searchResult := res.(map[interface{}]interface{})
+		Expect(searchResult["total_results"]).To(BeEquivalentTo(int64(2)))
+		Expect(searchResult["results"].([]interface{})[0].(map[interface{}]interface{})["id"]).To(BeEquivalentTo("doc2"))
```
Comment (chayim) [inline search_test.go:47]: This looks uncastable. can you try?

Comment (chayim) [reply search_test.go:47]: IMHO points to the need for helpers on the return

File: search_commands.go (hunk 10 above)
Comment (ofekshenawa) [reply search_commands.go:41]: Agree. However, this PR is a low-level RediSearch implementation. NewFTIndexWith<> functions will be in a future update to simplify common use cases.

File: search_commands.go (hunk 5 above)
Comment (ofekshenawa) [reply search_commands.go:44]: The `baseCmd` already converts `[]interface{}` into strings, so we don't need to implement this in the function. Also, we have arguments like `Score` or `MaxTextFields` that aren't strings, so changing it to `[]strings` wouldn't be feasible.

File: search_commands.go (hunk 4 above)
Comment (ofekshenawa) [reply search_commands.go:57]: Same as Prefix

File: search_commands.go (hunk 6 above)
Comment (ofekshenawa) [reply search_commands.go:74]: This is a low-level implementation of RedisSearch. We plan to add a high-level implementation for vector and other types in the future.

File: search_commands.go (hunk 9 above)
Comment (ofekshenawa) [reply search_commands.go:799]: fixed to match:
https://tip.golang.org/doc/comment#links

File: search_test.go (hunk 12 above)
Comment (ofekshenawa) [reply search_test.go:47]: In Go, handling unstable return types like in the FTSearch function is challenging due to its static typing. Creating a helper function to determine the return type isn’t feasible due to the dynamic nature of the return type and the infinite possible variations.

**Comment by jiangtao244** [comment]
how is this mr going？

**Review by vladvildanov — DISMISSED** [review]

**Review by vladvildanov — APPROVED** [review]

**Comment by quenbyako** [comment]
@ofekshenawa @vladvildanov @chayim hi guys, sorry for interrupting, we stucked with using RediSearch in golang, especially with RESP3 protocol: 

I found that RESP3 testing code was commented for some reason, so in production it doesn't work at all (incorrect response parsing) — any luck with supporting third version of protocol?

If there is no pull request which fixes that problem, I created a fork with **completed implementation** of supporting "new" version of response, should I create a pull request for that?

**Comment by vladvildanov** [comment]
@quenbyako Thank you for reaching us out! Sorry for the bad experience, but we had to rollback our latest release that brings a support for RediSearch, so this functionality is considered non-finished. We're planning to release it very soon, but on the first stage it will be **RESP2 only**, RESP3 support will be a part of the release after this.

Stay tuned for update in near future!

**Comment by quenbyako** [comment]
@vladvildanov Glad that you responded so quick, thank you 😊🙏

Regarding to features rollback: Was this lack of compatibility for RESP3 documented somewhere? Cause I spent yesterday nearly 5 hours to undestand what's happened, and it was hell, there is completely no info about this one...

I implemented [quick and dirty support for `ft.aggregate`](https://github.com/quenbyako/go-redis/compare/master..fix/redisearch-fix) command, please, take a look if you have some time, for our project redisearch is really important part, and I really want to collaborate with you guys on that feature addition, if that's possible 🤗

**Comment by vladvildanov** [comment]
@quenbyako The RESP3 will be partially compatible in near future and will be explained in the statement we prepare for community. We apologise that it forced you to dig into it, it was totally not planned to release it "that" way. We promise to properly communicate it with you 👌

**Comment by vladvildanov** [comment]
@quenbyako Join our discord community, there's a channel dedicated to go-redis as well https://discord.gg/zhvcaDMc

**Comment by quenbyako** [comment]
Hi @vladvildanov any update about this?

**Comment by vladvildanov** [comment]
@quenbyako Hey! Sorry for the delay, we're about to release Search support along with JSON in RESP2, will be included into next release

**Comment by sainad2222** [comment]
Hi @vladvildanov, Is there any ongoing issue that I can track for progress? Not able to join discord invite that was shared above too

**Comment by vladvildanov** [comment]
@sainad2222 Hey! Search support has been released and it's a part of 9.7.0

**Pull request merged**